### PR TITLE
Add Configurer

### PR DESCRIPTION
**Pull request by OlleLarsson**
`configure-es -> configurer`

The configurer script needs some more work.

### CODE REVIEW

File: helm/opendistro-es/templates/configurer/configurer-config.yml
Comment (lentzi90) [inline helm/opendistro-es/templates/configurer/configurer-config.yml]: These are not really templated, maybe skipt the `tpl`? Alternatively, make a single template that can be used for all of them. :slightly_smiling_face:

**Review by lentzi90 — COMMENTED** [review]
This looks good to me! :slightly_smiling_face: 
Should the configurer configmap also have the hook annotations to make its lifetime synchronized with the job?

**Comment by OlleLarsson** [comment]
> This looks good to me! 
> Should the configurer configmap also have the hook annotations to make its lifetime synchronized with the job?

Yeah, good catch, the configmap slippe my mind :slightly_smiling_face:

File: helm/opendistro-es/templates/configurer/configurer-config.yml
Comment (danielharr) [reply helm/opendistro-es/templates/configurer/configurer-config.yml]: Removed the `tpl` :slightly_smiling_face:

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (OlleLarsson) [inline helm/opendistro-es/files/configurer/configurer.sh]: Maybe we can add the parameter `create` to be able to decide wether to overwrite existing templates or not.
https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-templates.html#put-index-template-api-query-params

File: helm/opendistro-es/templates/configurer/configurer-config.yml
Comment (OlleLarsson) [inline helm/opendistro-es/templates/configurer/configurer-config.yml:12]: Maybe you can add the following annotations 
```
"helm.sh/hook": {{ .Values.configurer.helm.hook }}
"helm.sh/hook-weight": {{ .Values.configurer.helm.hookWeight | quote }}
"helm.sh/hook-delete-policy": {{ .Values.configurer.helm.deletePolicy }}
```
Then the configmap should follow the same lifecycle as the job :slightly_smiling_face:

File: helm/opendistro-es/files/configurer/kubeaudit_template.json
Comment (OlleLarsson) [inline helm/opendistro-es/files/configurer/kubeaudit_template.json]: I see the the indentation is off for all json files. Maybe we should only use an indentation of 2 spaces?
```suggestion
{
  "index_patterns": [
    "kubeaudit*"
  ],
  "settings": {
    "index.refresh_interval": "10s",
    "opendistro.index_state_management.policy_id": "kubeaudit",
    "opendistro.index_state_management.rollover_alias": "kubeaudit",
    "mapping.total_fields.limit": "1000"
  }
}
```

File: helm/opendistro-es/files/kibana-resources/README.md
Comment (OlleLarsson) [inline helm/opendistro-es/files/kibana-resources/README.md]:
```suggestion
- Index patterns for `kubernetes*`, `kubeaudit*`, and `other*` (`kubernetes*` is set as default)
```
I am unsure if it still tries to create the `kubecomponents*` index pattern :thinking: ?

File: helm/opendistro-es/values.yaml
Comment (OlleLarsson) [inline helm/opendistro-es/values.yaml:563]: Considering if we should add an option that enables or disables the inital index creation. Might come in handy in restore situations not to create the inital indexes.
```suggestion

  createIndices: true # or createIndexes, not sure which one to use

```

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (OlleLarsson) [inline helm/opendistro-es/files/configurer/configurer.sh:182]: Maybe we can split up the string into several lines? Or maybe we can read it from a file?

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (OlleLarsson) [inline helm/opendistro-es/files/configurer/configurer.sh:162]: Is it fine to run this function even if the roles already exist?

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (OlleLarsson) [inline helm/opendistro-es/files/configurer/configurer.sh:93]: Maybe we want to add a flag if we ever feel like we don't want to overwrite existing policies?

**Review by OlleLarsson — COMMENTED** [review]
Good work. I have added some comments and questions.

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (danielharr) [reply helm/opendistro-es/files/configurer/configurer.sh:182]: Reading from a file would probably be good!
The file tree could be organized more, the `configurer` folder could benefit from more sub-folders :slightly_smiling_face:

**Comment by OlleLarsson** [comment]
Just so that I am up to date.
If this job were to be run multiple times, the templates, policies, and kibana dashboards would be overwritten right?

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (danielharr) [reply helm/opendistro-es/files/configurer/configurer.sh]: Ah yes, it's available in opendistro as well
https://opendistro.github.io/for-elasticsearch-docs/docs/elasticsearch/rest-api-reference/#indicesput_template

**Comment by danielharr** [comment]
> Just so that I am up to date.
> If this job were to be run multiple times, the templates, policies, and kibana dashboards would be overwritten right?

Yes, that's the behavior I've been going for so far :slightly_smiling_face:

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (danielharr) [reply helm/opendistro-es/files/configurer/configurer.sh:162]: It will update them if they exist

File: helm/opendistro-es/files/kibana-resources/README.md
Comment (lentzi90) [reply helm/opendistro-es/files/kibana-resources/README.md]: Ah good point! I'll add an issue to purge `kubecomponents` everywhere :slightly_smiling_face:

File: helm/opendistro-es/templates/configurer/configurer-config.yml
Comment (danielharr) [reply helm/opendistro-es/templates/configurer/configurer-config.yml:12]: Added :slightly_smiling_face:

File: helm/opendistro-es/files/configurer/kubeaudit_template.json
Comment (danielharr) [reply helm/opendistro-es/files/configurer/kubeaudit_template.json]: Fixed :slightly_smiling_face:

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (danielharr) [reply helm/opendistro-es/files/configurer/configurer.sh:93]: This is in progress, but have to implement some logic for handling if we want to update/override the index-policy relationship.

EDIT:
A flag for overwriting existing policies has been added.
The relationship index<->policy will be handled as following:
If an index already has a policy, it will just be logged. Partial successes will be logged, to clarify if any indices were affected.
This way, old indices should not change policy, and if a new index somehow didn't get a policy, it will get one once the configurer is run.

File: helm/opendistro-es/values.yaml
Comment (danielharr) [reply helm/opendistro-es/values.yaml:563]: Added :slightly_smiling_face:

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (danielharr) [reply helm/opendistro-es/files/configurer/configurer.sh]: Added :slightly_smiling_face:

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (danielharr) [reply helm/opendistro-es/files/configurer/configurer.sh:182]: Reorganized the files used by configurer under the `files/configurer` folder

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (OlleLarsson) [inline helm/opendistro-es/files/configurer/configurer.sh]: I think that we can skip this step right here. The policy should be assigned to the indexes via the index templates that we create.
```suggestion
```

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (OlleLarsson) [inline helm/opendistro-es/files/configurer/configurer.sh]:
```suggestion
  echo "Creating initial indices"
```

File: helm/opendistro-es/values.yaml
Comment (OlleLarsson) [inline helm/opendistro-es/values.yaml]:
```suggestion
  ism:
```

**Comment by vanneback** [comment]
Have not looked through all the logic. But nice cleanup of this job @danielharr !

File: helm/opendistro-es/values.yaml
Comment (OlleLarsson) [inline helm/opendistro-es/values.yaml]: Maybe?
```suggestion
  # If true, configurer will overwrite existing templates.
  # See 'create' parameter:
  # https://opendistro.github.io/for-elasticsearch-docs/docs/elasticsearch/rest-api-reference/#indicesput_template
  overwriteTemplates: true
```

**Review by OlleLarsson — COMMENTED** [review]
I have added some more minor comments :slightly_smiling_face:

File: helm/opendistro-es/files/configurer/configurer.sh
Comment (danielharr) [reply helm/opendistro-es/files/configurer/configurer.sh]: Great! Committed the suggestion and appended some cleanup locally :slightly_smiling_face:

File: helm/opendistro-es/values.yaml
Comment (danielharr) [reply helm/opendistro-es/values.yaml]: Added + fixed templates

File: helm/opendistro-es/values.yaml
Comment (danielharr) [reply helm/opendistro-es/values.yaml]: This would be a negation of the API, which is why I implemented it as I did. So, pretty much to keep configuration similar to opendistro for those familiar to it.
I agree that `overwriteTemplates` looks better though.
What do you think?

Comment (OlleLarsson) [reply helm/opendistro-es/values.yaml]: Yeah, now that i read it in the documention it sort of sounds overly complicated.
If i get to choose I would go with the negation and keep `overwriteTemplates: true` :slightly_smiling_face:

Comment (danielharr) [reply helm/opendistro-es/values.yaml]: It added a requirement of checking that we actually supply `true` or `false`, but maybe that's handled by helm typing it as a boolean? Not entirely satisfied with how it ended up but not a big deal I suppose.

Comment (OlleLarsson) [reply helm/opendistro-es/values.yaml]: It's fine to revert if you want if, if the previous solution is cleaner. This is totaly fine with me :slightly_smiling_face:

Comment (danielharr) [reply helm/opendistro-es/values.yaml]: Both are fine I guess, can keep it unless anyone objects :slightly_smiling_face:

**Review by lentzi90 — APPROVED** [review]
Great job!
I have only one suggestion (and it may not be worth it, your call).

The `*_template.json` and `*_policy.json` files could be templated a bit more so that we only have one of each. We could then loop over index names for the configmap just as in the script.

It could look something like this:
```json
{
  "index_patterns": [
    "{{ $index }}*"
  ],
  "settings": {
    "index.refresh_interval": "10s",
    "opendistro.index_state_management.policy_id": "{{ $index }}",
    "opendistro.index_state_management.rollover_alias": "{{ $index }}",
    "mapping.total_fields.limit": "1000"
  }
}
```

To build the configmap I think something like this should work:

```
kind: ConfigMap
...
data:
{{- $files := .Files }}
  {{- range $index := tuple "kubeaudit" "kubernetes" "other" }}
  {{ $index }}_template.json: |-
    {{ $files.Get "files/configurer/index-templates/*.json" }}
  {{- end }}
```

I have pieced this together from the following docs:
- https://helm.sh/docs/chart_template_guide/accessing_files/
- https://helm.sh/docs/chart_template_guide/variables/
- https://helm.sh/docs/chart_template_guide/control_structures/

**Comment by OlleLarsson** [comment]
The tricky part with templating them in that way is that they essentially become the same. This might be what we want but it might as well not. For tempus we have set explicit mappings on the kubernetes template, we have increased the mapping limit etc. And I mean we might not want to treat the kubernetes index exactly the same as other and kubeaudit.

I am fine with templating as Lennart suggested, but i would ideally also have the capability at some point in the _future_ to specifify additional template and policy specification through helm values

**Comment by lentzi90** [comment]
> I am fine with templating as Lennart suggested, but i would ideally also have the capability at some point in the _future_ to specifify additional template and policy specification through helm values

Agreed! I think templating a little bit now could make it easier to add more later, but I don't have a strong opinion on this right now. We can keep it as is or template, both should be fine for now. :slightly_smiling_face:

**Comment by danielharr** [comment]
> > I am fine with templating as Lennart suggested, but i would ideally also have the capability at some point in the _future_ to specifify additional template and policy specification through helm values
> 
> Agreed! I think templating a little bit now could make it easier to add more later, but I don't have a strong opinion on this right now. We can keep it as is or template, both should be fine for now.

I was wondering about this, and did not change anything since I don't know how likely it is that the policies will remain identical.
I'm not sure if it's possible to "extend" the template for just one of the policies either, without removing it from the loop and using a separate template.

If it's likely that at least two policies will remain identical, I'm all for making it more DRY :slightly_smiling_face:

**Comment by OlleLarsson** [comment]
They will probably remain identical at least in the nearest future.

**Comment by danielharr** [comment]
I've made an attempt to make this work, but couldn't find a solution to getting the correct scope with `tpl` and the `range` variable (for example `{{ $index }}`).

To my understanding `tpl` requires access to the root scope, which is fine:
```
{{- $files := .Files }}
  {{- range $policy := tuple "kubeaudit" "kubernetes" "other" }}
  {{ $policy }}_template.json: |-
    {{ tpl ($files.Get "files/configurer/ism-policies/policy_template.json") $ }}
  {{- end }}
```
however `$policy` is undefined. Perhaps the `$policy` value could be added to the values dict somehow, but I feels like it's getting too complex.

_Doesn't work, but maybe something like this:_
```
{{- $files := .Files }}
  {{- range $policy := tuple "kubeaudit" "kubernetes" "other" }}
  {{ $policy }}_template.json: |-
    {{ tpl ($files.Get "files/configurer/ism-policies/policy_template.json") (dict "root" $.Values "policy" $policy ) }}
  {{- end }}
```

Feel free to give it a go if you like, otherwise I'm fine merging it as it is :slightly_smiling_face:

**Comment by OlleLarsson** [comment]
I'll hold off on merging this as I want to first try it out, but i agree this looks good the way it is :slightly_smiling_face:

**Comment by OlleLarsson** [comment]
Okay, I have now tested this and i just found out that we can't use the same `hook-weight` for the configmap and the job. If we do then the job will be created before the cm, which means that the job cannot start because the cm does not exist. The cm won't be created until the job has completed... which it never will.

So, either we set different fixed values on the two resources or we can do something like:
```
configurer:
  enabled: false

  helm:
    hook: post-install
    deletePolicy: before-hook-creation,hook-failed
    cm:
      hookWeight: -10
    job:
      hookWeight: -5
```

Also, right now the job is only run on first install. We want it run for upgrades as well, right? So basicially `configurer.helm.hook: post-install, post-upgrade`.

What do you guys think?

**Comment by lentzi90** [comment]
> So, either we set different fixed values on the two resources or we can do something like:
> 
> ```
> configurer:
>   enabled: false
> 
>   helm:
>     hook: post-install
>     deletePolicy: before-hook-creation,hook-failed
>     cm:
>       hookWeight: -10
>     job:
>       hookWeight: -5
> ```
> 
> What do you guys think?

I think fixed values makes sense in this case since this cannot work if the values are changed.

**Comment by OlleLarsson** [comment]
> I think fixed values makes sense in this case since this cannot work if the values are changed.

I agree!

File: helm/opendistro-es/templates/configurer/configurer.yaml
Comment (lentzi90) [inline helm/opendistro-es/templates/configurer/configurer.yaml:12]: Can you add a comment so we don't forget that this is "linked" to the configmap?

Comment (OlleLarsson) [reply helm/opendistro-es/templates/configurer/configurer.yaml:12]: Do you mean like?
```
# Use higher value so that the configmap is created before this job
"helm.sh/hook-weight": "-5"
```

Comment (lentzi90) [reply helm/opendistro-es/templates/configurer/configurer.yaml:12]: Yes, perfect!

Comment (OlleLarsson) [reply helm/opendistro-es/templates/configurer/configurer.yaml:12]: Done

**Comment by OlleLarsson** [comment]
Alright, I have tested this now and I am happy with how it works.
If there are no more comments i will squash and merge this :slightly_smiling_face:
Great work Daniel!

**Comment by OlleLarsson** [comment]
This is something that we can decide upon later but right now if we delete the release the configmap and job will be left behind.
I just wanted to have this mentioned for future reference (so that i don't forget it :P ).

**Pull request merged**
